### PR TITLE
[Focus] Add run key command to dump the focus tree

### PR DESCRIPTION
**Pull request by loic-sharma**
Adds a Flutter run key to dump the focus tree: once the app is launched using `flutter run`, pressing `f` dumps the focus tree. This aligns with how pressing `w` dumps the widget hierarchy and `t` dumps the render tree.

This makes it easier to debug focus issues like:

* What is focused currently?
* Why is tabbing moving my focus unexpectedly?
* Why isn't my shortcut's action running? 

Example output:

```
> flutter run lib\empty.dart -d windows
Launching lib\empty.dart on Windows in debug mode...
Building Windows application...                                     5.4s
✓  Built build\windows\runner\Debug\ui.exe.
Syncing files to device Windows...                                  66ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h List all available interactive commands.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).

A Dart VM Service on Windows is available at: http://127.0.0.1:49880/aAvLj-OYB2c=/
The Flutter DevTools debugger and profiler on Windows is available at:
http://127.0.0.1:9102?uri=http://127.0.0.1:49880/aAvLj-OYB2c=/
```

Then when you press `f`:

```

FocusManager#59e1c
 └─rootScope: FocusScopeNode#8b67c(Root Focus Scope)

```

Part of https://github.com/flutter/flutter/issues/118536

## Screen recording

dump_focus_tree

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/test/foundation/service_extensions_test.dart [hunk 1]
```diff
@@ -218,6 +218,14 @@ void main() {
     });
   });
 
+  test('Service extensions - debugDumpFocusTree', () async {
+    final Map<String, dynamic> result = await binding.testExtension(WidgetsServiceExtensions.debugDumpFocusTree.name, <String, String>{});
+
+    expect(result, <String, dynamic>{
+      'data': matches('TestServiceExtensionsBinding - DEBUG MODE\n<no tree currently mounted>'),
```
Comment (loic-sharma) [inline packages/flutter/test/foundation/service_extensions_test.dart:225]: TODO

File: packages/flutter_tools/bin/flutter_tools.dart [hunk 2]
```diff
@@ -2,8 +2,15 @@
...
 
+import 'dart:io';
+
 import 'package:flutter_tools/executable.dart' as executable;
 
 void main(List<String> args) {
+  if (args.isNotEmpty && args[0] == 'run') {
+    Directory.current = r'C:\Code\f\flutter\dev\integration_tests\ui\';
+    args = const <String>['run', '-d', 'windows', r'lib\empty.dart'];
+  }
```
Comment (loic-sharma) [inline packages/flutter_tools/bin/flutter_tools.dart:13]: Undo

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

Looks great, thanks!

**Review by zanderso — APPROVED** [review]
LGTM

**Pull request merged**